src/boros_impl/op/read.c

```diff
@@ -0,0 +1,99 @@
+/* This source file is part of the boros project. */
+/* SPDX-License-Identifier: ISC */
+
+#include "op/read.h"
+
+#include <liburing.h>
+
+#include "bytesobject.h"
+#include "pyerrors.h"
+#include "util/python.h"
+#include "module.h"
+
+static void read_prepare(PyObject *self, struct io_uring_sqe *sqe) {
+    ReadOperation *op = (ReadOperation *)self;
+
+    char *buf;
+
+    buf = PyBytes_AsString(op->buf);
+
+    io_uring_prep_read(sqe, op->fd, buf, op->nbytes, op->offset);
+}
+
+static void read_complete(PyObject *self, struct io_uring_cqe *cqe) {
+    Operation *op = (Operation *)self;
+    outcome_capture(&op->outcome, PyLong_FromLong(cqe->res));
```

Comment:
This logic is incorrect. `cqe->res < 0` indicates an error, so you want to call `outcome_capture_errno(-cqe->res)` in that case. Otherwise, call `_PyBytes_Resize(&op->buf, cqe->res)` and capture the buffer into the otucome.